server/service/activities.go

```diff
@@ -43,8 +53,78 @@ func (svc *Service) ListActivities(ctx context.Context, opt fleet.ListActivities
 	return svc.ds.ListActivities(ctx, opt)
 }
 
+type ActivityWebhookPayload struct {
+	Timestamp     time.Time        `json:"timestamp"`
+	ActorFullName *string          `json:"actor_full_name"`
+	ActorID       *uint            `json:"actor_id"`
+	ActorEmail    *string          `json:"actor_email"`
+	Type          string           `json:"type"`
+	Details       *json.RawMessage `json:"details"`
+}
+
 func (svc *Service) NewActivity(ctx context.Context, user *fleet.User, activity fleet.ActivityDetails) error {
-	return svc.ds.NewActivity(ctx, user, activity)
+	return newActivity(ctx, user, activity, svc.ds, svc.logger)
+}
+
+func newActivity(ctx context.Context, user *fleet.User, activity fleet.ActivityDetails, ds fleet.Datastore, logger kitlog.Logger) error {
+	appConfig, err := ds.AppConfig(ctx)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "get app config")
+	}
+
+	detailsBytes, err := json.Marshal(activity)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "marshaling activity details")
+	}
+	timestamp := time.Now()
+
+	if appConfig.WebhookSettings.ActivitiesWebhook.Enable {
+		webhookURL := appConfig.WebhookSettings.ActivitiesWebhook.DestinationURL
+		var userID *uint
+		var userName *string
+		var userEmail *string
+		if user != nil {
+			userID = &user.ID
+			userName = &user.Name
+			userEmail = &user.Email
+		}
+		activityType := activity.ActivityName()
+		go func() {
+			retryStrategy := backoff.NewExponentialBackOff()
+			retryStrategy.MaxElapsedTime = 30 * time.Minute
+			err := backoff.Retry(
+				func() error {
+					if err := server.PostJSONWithTimeout(
+						context.Background(), webhookURL, &ActivityWebhookPayload{
+							Timestamp:     timestamp,
+							ActorFullName: userName,
+							ActorID:       userID,
+							ActorEmail:    userEmail,
+							Type:          activityType,
+							Details:       (*json.RawMessage)(&detailsBytes),
+						},
+					); err != nil {
+						var statusCoder kithttp.StatusCoder
+						if errors.As(err, &statusCoder) && statusCoder.StatusCode() == http.StatusTooManyRequests {
+							level.Debug(logger).Log("msg", "fire activity webhook", "err", err)
+							return err
+						}
+						return backoff.Permanent(err)
+					}
+					return nil
+				}, retryStrategy,
+			)
+			if err != nil {
+				level.Error(logger).Log(
+					"msg", fmt.Sprintf("fire activity webhook to %s", server.MaskSecretURLParams(webhookURL)), "err",
+					server.MaskURLError(err).Error(),
+				)
+			}
+		}()
+	}
+	// We update the context to indicate that we processed the webhook.
+	ctx = context.WithValue(ctx, fleet.ActivityWebhookContextKey, true)
```

Comment:
i'm a bit concerned that adding this to the context isn't idomatic in Go:
 https://pkg.go.dev/context#:~:text=key%20is%20val.-,Use%20context%20Values%20only%20for%20request%2Dscoped%20data%20that%20transits,collisions%20between%20packages%20using%20context.

Reply:
Good catch. I created a new type.

Reply:
sorry, what I meant was that it seems like we're using context to pass a boolean parameter

```
Use context Values only for request-scoped data that transits processes and APIs, not for passing optional parameters to functions.
```

Reply:
Yes, that's true. I think this is a valid use of context -- we are indicating whether the request has processed the activity webhook. So, we're tracking the state of request processing.